### ISSUE TITLE
Inappropriately handled Postgres message

### ISSUE THREAD

**Issue by isaacsanders** [elixir-ecto/postgrex#405]
Erlang Version: `20.3.7`
Elixir Version: `1.6.5`
Postgres Version: `10.3`
Library Version: `0.13.5`
Exception:
```
12:33:55.064 [error] GenServer #PID<0.397.0> terminating
** (DBConnection.ConnectionError) client #PID<0.498.0> stopped: ** (FunctionClauseError) no function clause matching in Postgrex.Protocol.handle_msg/3
    (postgrex) lib/postgrex/protocol.ex:2062: Postgrex.Protocol.handle_msg(%Postgrex.Protocol{buffer: nil, connection_id: 8752, connection_key: -1565263751, null: nil, parameters: #Reference<0.673783574.3612344324.52694>, peer: {{127, 0, 0, 1}, 5432}, postgres: :idle, queries: #Reference<0.673783574.3612475396.52691>, sock: {:gen_tcp, #Port<0.9064>}, timeout: 15000, transactions: :naive, types: {MyApp.PostgresTypes, #Reference<0.673783574.3612475396.52652>}}, %{mode: :transaction, notify: #Function<32.52153523/2 in Postgrex.Protocol.notify/1>, sync: :flushed_sync}, {:msg_error, [severity: "ERROR", unknown: "ERROR", code: "23503", message: "insert or update on table \"children\" violates foreign key constraint \"children_parent_id_fkey\"", detail: "Key (parent_id)=(213143) is not present in table \"parents\".", schema: "public", table: "children", constraint: "children_parent_id_fkey", file: "ri_triggers.c", line: "3321", routine: "ri_ReportViolation"]})
    (postgrex) lib/postgrex/protocol.ex:2183: Postgrex.Protocol.do_sync_recv/4
    (db_connection) lib/db_connection.ex:966: DBConnection.handle/4
    (db_connection) lib/db_connection.ex:1086: DBConnection.describe_run/5
    (db_connection) lib/db_connection.ex:1150: anonymous fn/4 in DBConnection.run_meter/5
    (db_connection) lib/db_connection.ex:1207: DBConnection.run_begin/3
    (db_connection) lib/db_connection.ex:592: DBConnection.prepare_execute/4
    (ecto) lib/ecto/adapters/postgres/connection.ex:86: Ecto.Adapters.Postgres.Connection.execute/4

    (db_connection) lib/db_connection/connection.ex:258: DBConnection.Connection.handle_cast/2
    (connection) lib/connection.ex:488: Connection.handle_call/3
    (stdlib) gen_server.erl:636: :gen_server.try_handle_call/4
    (stdlib) gen_server.erl:665: :gen_server.handle_msg/6
    (stdlib) proc_lib.erl:247: :proc_lib.init_p_do_apply/3
```

Situation: I am inserting data into a table via `Ecto.Repo.insert_all`, the insertions are happening in parallel, but I see in the logs that the parent has been inserted beforehand. I realize this might just be a race condition, but I think the exception handling in Postgrex is wrong. `Postgrex.Protocol.handle_msg/3` should handle more of the messages, or be renamed, because it only handles 3 of the message record types, as far as I found.

Expected: I would expect to get a non-exception response.

**Comment by josevalim** [comment]
Thanks for the report. Can you provide a minimal app that reproduces the issue? Otherwise we might spend a lot of time trying to reproduce the issue and we are time constrained.

**Comment by fishcakez** [comment]
@isaacsanders can you try with https://github.com/elixir-ecto/postgrex/commit/940c1126a68ad0df065c79980f0fc776ba10633a. I believe we fixed this in master by that point and that commit will still be backwards compatible enough for your other deps.

**Comment by fishcakez** [comment]
Oh with postgres 10 will need https://github.com/elixir-ecto/postgrex/commit/b93e729c96ba5be5d95f41f370c66d80e8ea94e1 cherrypicked on top.

**Comment by josevalim** [comment]
Ping.

**Comment by isaacsanders** [comment]
I don't know the combination of things that occur to cause this error, but I can try @fishcakez's solution. I will let y'all know how it goes.

**Comment by josevalim** [comment]
We are releasing postgrex 0.14-rc and a new ecto version should be out this week. We believe this is fixed so we will go ahead and close it. If the issue persists, please comment here or open up a new issue with a mechanism to reproduce it. thanks!
